src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.Accessor.cs

```diff
@@ -68,35 +68,96 @@ public Accessor(SQLitePersistentStorage storage)
 
             protected abstract Table Table { get; }
 
-            protected abstract bool TryGetDatabaseId(SqlConnection connection, TKey key, out TDatabaseId dataId);
+            /// <summary>
+            /// Gets the internal sqlite db-id (effectively the row-id for the doc or proj table, or just the string-id
+            /// for the solution table) for the provided caller key.  This db-id will be looked up and returned if a
+            /// mapping already exists for it in the db.  Otherwise, a guaranteed unique id will be created for it and
+            /// stored in the db for the future.  This allows all associated data to be cheaply associated with the 
+            /// simple ID, avoiding lots of db bloat if we used the full <paramref name="key"/> in numerous places.
+            /// </summary>
+            /// <param name="allowWrite">Whether or not the caller owns the write lock and thus is ok with the DB id
+            /// being generated and stored for this component key when it currently does not exist.  If <see
+            /// langword="false"/> then failing to find the key will result in <see langword="false"/> being returned.
+            /// </param>
+            protected abstract bool TryGetDatabaseId(SqlConnection connection, TKey key, bool allowWrite, out TDatabaseId dataId);
             protected abstract void BindFirstParameter(SqlStatement statement, TDatabaseId dataId);
             protected abstract TWriteQueueKey GetWriteQueueKey(TKey key);
             protected abstract bool TryGetRowId(SqlConnection connection, Database database, TDatabaseId dataId, out long rowId);
 
             [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/36114", AllowCaptures = false)]
-            public Task<bool> ChecksumMatchesAsync(TKey key, Checksum checksum, CancellationToken cancellationToken)
-                => Storage.PerformReadAsync(
-                    static t => t.self.ChecksumMatches(t.key, t.checksum, t.cancellationToken),
-                    (self: this, key, checksum, cancellationToken), cancellationToken);
+            public async Task<bool> ChecksumMatchesAsync(TKey key, Checksum checksum, CancellationToken cancellationToken)
+            {
+                var (succeeded, dataId) = await TryGetDatabaseIdAsync(key, cancellationToken).ConfigureAwait(false);
+                if (!succeeded)
+                    return false;
+
+                return await Storage.PerformReadAsync(
+                    static t => t.self.ChecksumMatches(t.dataId, t.checksum, t.cancellationToken),
```

Comment:
❓ @CyrusNajmabadi is it possible to push the `TryGetDatabaseIdAsync` call into `PerformReadAsync` (and then one step further into `PerformTaskAsync`) to avoid creating new async state machines on this path? If this works, it would be the easiest way to ensure we do not regress performance.

Reply:
we coudl.  though we'd still need to do two calls at the end of the day as one call may be able to run on the concurrent queue, but we need to transfer over to hte exclusive queue if it cannot proceed.